src/babashka/process.cljc

```diff
@@ -178,17 +178,28 @@
       (str/includes? "windows")))
 
 (defn- -program-resolver [{:keys [program dir]}]
-  ;; this should make life easier and not cause any bugs that weren't there previously
-  ;; on exception we just return the program as is
-  (try
-    (if (fs/relative? program)
-      (if-let [f (fs/which (if dir
-                             (-> (fs/file dir program) fs/absolutize)
-                             program))]
-        (str f)
-        program)
-      program)
-    (catch Throwable _ program)))
+  ;; Used by default:
+  ;; - for all program resolution when running on Windows
+  ;; - by `exec` on all OSes
+  ;; Default ProcessBuilder behaviour on macOS/Linux does this work naturally
+  ;; and is therefore unneeded.
+  ;;
+  ;; ProcessBuilder program resolution on Windows does not entirely match
+  ;; behaviour of CMD Shell nor PowerShell.
+  ;; Here we adapt resolution to match PowerShell (with the exclusion of resolving
+  ;; .ps1 scripts) which follows the same strategy as macOS/Linux.
+  (if-let [resolved (cond
+                      (fs/absolute? program)
+                      (fs/which program) ;; to resolve any extensions
+
+                      (fs/parent program)
+                      (fs/which (fs/file (or dir (fs/cwd))
```

Comment:
This works, but it would also work when written like this, which feels less magic to me:

```
(fs/which (if dir (fs/file dir program) program))
```

which leaves the `program` argument unaltered if you provide no dir

Reply:
We need to resolve Windows filename extensions even if no dir is provided. 
[ProcessBuilder only resolves .exe](https://github.com/lread/info-process-builder/blob/main/README.adoc#how-does-a-resolve-from-cwd-1).

Reply:
True, so then it would become:

``` clojure
(if dir
  (fs/which (fs/file dir program))
  (fs/which program))
```

which feels a little bit more explicit to me then using `fs/cwd` which alters the program path with an extra cwd entry

Reply:
Now I'm questioning this too... I probably include cwd for some reason... but can explore.